src/boot/efi/initrd.c

```diff
@@ -88,19 +88,28 @@ EFI_STATUS initrd_register(
            where as InstallMultipleProtocolInterfaces only matches identical DevicePaths.
          */
         err = BS->LocateDevicePath(MAKE_GUID_PTR(EFI_LOAD_FILE2_PROTOCOL), &dp, &handle);
-        if (err != EFI_NOT_FOUND) /* InitrdMedia is already registered */
-                return EFI_ALREADY_STARTED;
+        if (err == EFI_SUCCESS && dp->Type == END_DEVICE_PATH_TYPE) {
+                /* InitrdMedia is already registered */
+                *ret_initrd_handle = handle;
+                initrd_unregister ( handle );
```

Comment:
Can't we just uninstall the device path protocol from the other handle here, and proceed as usual?

Reply:
The problem with doing that is that it may leave some other piece of code holding a stale `EFI_HANDLE` value for a handle that no longer exists.  This feels risky to me.

Reply:
The handle will exist until the last protocol it carries is uninstalled.

The handle could carry other protocols that the provider may use in combination with the LoadFile2 protocol, so uninstalling that seems riskier to me than uninstalling the purely informational DevicePath protocol which is only used for locating the handle and nothing else.

Reply:
Ah, you mean to leave the `EFI_FILE_LOAD2_PROTOCOL` installed but uninstall the `EFI_DEVICE_PATH_PROTOCOL` handle instead?  Then yes, that should also work, although I haven't tested it.

Reply:
@ardbiesheuvel I'm happy to rework this PR to do the uninstallation of `EFI_DEVICE_PATH_PROTOCOL` instead of `EFI_FILE_LOAD2_PROTOCOL` if you like.

If so, it would probably also be worth documenting that this is the "correct" approach to take in the comments in `drivers/firmware/efi/libstub/efi-stub-helper.c`, updating `OvmfPkg/LinuxInitrdDynamicShellCommand/LinuxInitrdDynamicShellCommand.c` to support it, and changing iPXE to take this approach for consistency.

I'm happy to put together those assorted patches if that's your preferred approach - let me know.

Reply:
You could install a dummy protocol on the device path handle and then uninstall the device path protocol on that handle. That way the handle will stay valid and the next loader in the chain can register its own initrd device path (on a new handle). If the loader returns you reverse the process and the parent loader should be none the wiser and can itself properly clean up.

And this should be done in iPXE as it is the one consuming the protocol and therefore really is the one responsible for cleaning things up for the next one in the bootloader chain.

Reply:
> You could install a dummy protocol on the device path handle and then uninstall the device path protocol on that handle. That way the handle will stay valid and the next loader in the chain can register its own initrd device path (on a new handle). If the loader returns you reverse the process and the parent loader should be none the wiser and can itself properly clean up.

So basically exactly what iPXE already does, but choosing to break the symmetry on the other protocol of the pair.

> And this should be done in iPXE as it is the one consuming the protocol and therefore really is the one responsible for cleaning things up for the next one in the bootloader chain.

Other than requiring the ability to reverse time, this is a practical suggestion.  :slightly_smiling_face:

Reply:
> So basically exactly what iPXE already does, but choosing to break the symmetry on the other protocol of the pair.

If iPXE did that, you wouldn't be complaining here about things you broke. Gee, just fix your stuff already

Reply:
> > So basically exactly what iPXE already does, but choosing to break the symmetry on the other protocol of the pair.
> 
> If iPXE did that, you wouldn't be complaining here about things you broke. Gee, just fix your stuff already

iPXE is the component that does already support being part of a multi-bootloader chain.  The systemd EFI stub does not.  I could add some sarcastic workarounds to iPXE to allow for systemd bugs, but if possible then I'd prefer to just fix systemd.

> > Other than requiring the ability to reverse time, this is a practical suggestion.
> 
> Then please do explain to me how "fixing" the systemd side would magically not require your time travel device. I cannot follow that logic.

Sure.  In chronological order:
1. iPXE is used to download some files
2. iPXE is asked by the user to boot one of the downloaded files
3. iPXE performs format detection and determines that the selected boot file is a UEFI executable
4. iPXE uses `LoadImage()` and `StartImage()` to invoke the already-downloaded UEFI executable.
   * iPXE provides the loaded image with a virtual filesystem via `EFI_SIMPLE_FILE_SYSTEM_PROTOCOL` to allow it to access the other downloaded files.
   * The virtual filesystem also contains a special file `initrd.magic` which is basically the concatenation of all other downloaded files (in order of download, with appropriate initrd padding, and with the option to generate CPIO headers so that downloaded standalone files can be made to appear at a user-defined point within the resulting filesystem). This can be consumed by a kernel using the command-line option `initrd=initrd.magic`.
   * iPXE provides an `EFI_LOAD_FILE2_PROTOCOL` instance that gives access to the same `initrd.magic` file content.  This protocol instance is installed on a handle that has the `LINUX_INITRD_MEDIA_GUID` device path so that the kernel's EFI stub can locate and use it automatically, with no `initrd=initrd.magic` option required.  iPXE takes care to correctly handle the assorted subcases:
     * No handle with `LINUX_INITRD_MEDIA_GUID` exists yet: iPXE will create the handle and install the device path
     * A handle with `LINUX_INITRD_MEDIA_GUID` already exists: iPXE will locate the handle, uninstall the existing `EFI_LOAD_FILE2_PROTOCOL` instance, and install its own.

So far, so generic.  Let's now consider what the invoked UEFI executable could turn out to be:

5(a). The invoked UEFI executable happens to be a current Linux kernel.  It locates the initrd provided by iPXE via the `LINUX_INITRD_MEDIA_GUID` device path and retrieves it via the `EFI_LOAD_FILE2_PROTOCOL` instance and everything is happy in the world.

5(b). The invoked UEFI executable happens to be an older Linux kernel.  It locates the initrd provided by iPXE via the `initrd=initrd.magic` command-line option and retrieves it via the `EFI_SIMPLE_FILE_SYSTEM_PROTOCOL` instance and everything is happy in the world.

5(c). The invoked UEFI executable happens to be the UEFI shell.  It maps the `EFI_SIMPLE_FILE_SYSTEM_PROTOCOL` as `fs0:` and everything is happy in the world.

5(d). The invoked UEFI executable happens to be `wimboot`.  It retrieves the `boot.wim` and any other files required for the WinPE boot via the `EFI_SIMPLE_FILE_SYSTEM_PROTOCOL` instance and everything is happy in the world.

5(e). The invoked UEFI executable happens to be a Linux UKI **with this bugfix applied**.  The UKI EFI stub locates the existing `LINUX_INITRD_MEDIA_GUID` handle, removes the existing `EFI_LOAD_FILE2_PROTOCOL` instance, installs its own `EFI_LOAD_FILE2_PROTOCOL` instance to provide access to the initrd embedded inside the UKI, passes control to the kernel embedded inside the UKI, and everything is happy in the world.

5(f). The invoked UEFI executable happens to be a Linux UKI with a very similar bugfix applied as per https://github.com/systemd/systemd/pull/26735#discussion_r1131368943.  The UKI EFI stub locates the existing `LINUX_INITRD_MEDIA_GUID` handle, removes the existing `EFI_DEVICE_PATH_PROTOCOL` instance, creates a new handle with the `LINUX_INITRD_MEDIA_GUID` device path and its own `EFI_LOAD_FILE2_PROTOCOL` instance to provide access to the initrd embedded inside the UKI, passes control to the kernel embedded inside the UKI, and everything is happy in the world.

5(g). The invoked UEFI executable happens to be a Linux UKI built from the current systemd code **without a bugfix applied**.  The UKI EFI stub fails to create a handle with the `LINUX_INITRD_MEDIA_GUID` device path, reports an error, leaves the user unable to boot, and the world is no longer a happy place.

As you can see, no time travel involved in any of the above.  (Your suggested fix that iPXE should perform the cleanup required to make case 5(g) work would require some way for iPXE to temporarily regain control at the point that the systemd EFI stub calls `InstallMultipleProtocolInterfaces` to create its new handle.)